File: backend/src/main/java/com/accesodatos/entity/Product.java
package com.accesodatos.entity;

import java.util.ArrayList;
import java.util.List;

import com.accesodatos.entity.enums.ProductState;
import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Entity
@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(exclude = "buys")
@ToString(exclude = "buys")
@Table(name = "products")
public class Product {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "product_id")
	private Long productId;
	
	private String productName;
	
	private String productImage;
	
	private int price;
	
	@Enumerated(EnumType.STRING)
	private ProductState state;
	
	@OneToMany(
			mappedBy = "product",
			cascade = CascadeType.ALL,
			orphanRemoval = true
			)
	@JsonManagedReference("product-purchases")
	private List<Purchase> buys = new ArrayList<>();	
}

File: frontend/src/types/Product.ts
import productState from "./enums/ProductState";

type Product = {
  productId: number;
  productName: string;
  productImage: string;
  state?: productState;
  price: number;
};

type ProductRequest = {
  productName: string;
  productImage: string;
  state: productState;
  price: number;
};

interface ProductCartItemResponseDto {
  productId: number;
  productName: string;
  productImage: string;
  price: number;
}

type ProductResponse = {
  productId: number;
  productName: string;
  productImage: string;
  state?: productState;
  buys?: number;
  price: number;
};

type ProductPurchaseResponseDto = {
  productId: string;
  productName: string;
  productImage: string;
  price: number;
};

export type {
  Product,
  ProductPurchaseResponseDto,
  ProductRequest,
  ProductResponse,
  ProductCartItemResponseDto,
};
